platform-sdk/docs/proposals/consensus-layer/Consensus-Layer.md

```diff
@@ -0,0 +1,438 @@
+"hop", but allows us to identify dishonest gossip peers and discipline them accordingly.
+
+During execution, for all nodes that are online and able to keep up, events are received "live" and processed
+immediately and re-gossiped. However, if a node is offline and then comes back online, or is starting back up after
+reconnect, it may be missing events. In this case, the node will need to ask its peers for any events it is missing.
+
+For this reason, every honest node needs to buffer some events, so when its peer asks it for events, it is able to
+give them. Events can be "non-ancient", "ancient", or "expired". Although the definition for these is subject to
+configuration, we will here use some example values. Non-ancient events are those events that are either not yet
+in a consensus round, or are in one of the last 26 consensus rounds. Expired events are those that are in a round
+that is older than the last most recent 500 consensus rounds. And ancient events are those that are in a round between
+26 and 500.
+
+The Gossip module must cache all non-expired events. Note that it must be a fixed number of rounds that defines
+"expired", because otherwise it is possible for a quiescent network to never recover (see the documentation for
+the tipset algorithm, and other-parent selection, for specifics on why this is true). This number must also be large
+enough in case rounds/sec increases dramatically, because reconnect requires a certain amount of time (not rounds) for
+catchup. As long as this number is large enough to meet th requirements regardless of the rounds/sec, and as long as it
+is fixed in number of rounds, then it will work.
+
+#### Peer Discipline
+
+If a peer is misbehaving, the Gossip module will notify the Sheriff module that one of its peers is misbehaving. For
+example, if a peer is not responding to requests, even after repeated attempts to reconnect with it, it may be "bad".
+Or if the peer is sending events that exceed an acceptable rate, or exceed an acceptable side, then it is "bad". Or if
+the events it sends cannot be parsed, or are signed incorrectly, or in other ways fail validation, then it is "bad".
+There may be additional rules by the Gossip module or others (such as Event Intake detecting forking) that could lead to
+a peer being marked as "bad". A "bad" node may be dishonest, or it may be broken. The two cases are indistinguishable,
+so all such nodes are treated equally.
+
+If the Sheriff decides that the peer should be penalized, then it will instruct the Gossip module to "shun" that peer.
+"Shunning" is a unilateral behavior that one node can take towards another, where it terminates the connection and
+refuses to work further with that node. If the Sheriff decides to welcome a peer back into the fold, it can instruct the
+Gossip module to "welcome" the peer back.
+
+#### Falling Behind
+
+It may be that during gossip, the Gossip module finds that it cannot continue. Perhaps all connections have failed, or
```

Comment:
A heads up, the current fallen behind algorithm is not algorithmically airtight, especially in typologies where the number of peer connections is low. The correct way to do this is to use state proofs to prove that a peer is sufficiently far behind the current network consensus.